content/program-security.md

```diff
@@ -0,0 +1,741 @@
+    // 0
+    /// UnitializedAccount
+    #[error("Account not initialized yet")]
+    UnitializedAccount,
+    // 1
+    // Incorrect PDA
+    #[error("PDA derived does not equal PDA passed in")]
+    InvalidPDA,
+    // 2
+    // Data is too long
+    #[error("Input data exceeds max length")]
+    InvalidDataLength,
+    //3
+    // Rating is too high
+    #[error("Rating cannot exceed 5")]
+    RatingTooHigh,
+}
+
+impl From<ReviewError> for ProgramError {
+    fn from(e: ReviewError) -> Self {
+        ProgramError::Custom(e as u32)
+    }
+}
+```
+
+Then bring in the following crate to processor.rs like so
+
+```rust
+// inside processor.rs
+use crate::error::ReviewError;
+```
+
+### 4.Account Validation Checks
+
+Now you can refactor the `is_initialized` validation check to return the new `UnitializedAccount` error type that we just created.
+
+```rust
+msg!("checking if movie account is initialized");
+if !account_data.is_initialized() {
+  msg!("Account is not initialized");
+  return Err(ReviewError::UnitializedAccount.into());
+}
+```
+
+Our next validation check will be on the pda itself. We will derive the pda ourselves with the initializer’s public key and the title that is currently stored in the movie review. The reason we’re doing this is because you should only be able to update a review that was created by you and since the way the pdas were derived in the first place is with the initializer’s key and the title, this is an implicit check that the `initializer` passed in here did indeed create this review. If the `initializer` provided here is not the same public key that was used to derive this pda, then the pda that we derive program side will not equal the pda that was passed in. We’ll also return the second custom error type that we defined in [error.rs](http://error.rs) if they don’t match.
+
+```rust
+// Derive PDA and check that it matches client
+let (pda, _bump_seed) = Pubkey::find_program_address(&[initializer.key.as_ref(), account_data.title.as_bytes().as_ref(),], program_id);
+
+if pda != *pda_account.key {
+    msg!("Invalid seeds for PDA");
+	// return custom error
+    return Err(ReviewError::InvalidPDA.into())
+}
+```
+
+The last *account* validation will be to check that the initializer account signed this transaction. This ensures without a doubt that the creator of this review wants it to be updated since a signature cannot be provided without them using their private key and is generally just good practice.
+
+```rust
+if !initializer.is_signer {
+  msg!("Missing required signature");
+  return Err(ProgramError::MissingRequiredSignature)
+}
+```
+
+### 5. Instruction Data Validation
+
+There is a scenario that we have not accounted for yet. Remember when a review account is created, a specific amount of space is allocated based on the length of the data that it’s going to be storing with this line here
+`*let* account_len: *usize* = 1 + 1 + (4 + *title*.len()) + (4 + *description*.len());`
```

Comment:
@jamesrp13 seems odd to newline separate this code text... on my end that makes it render as just a part of the sentence continuously.  I suggest either 

sugg1: remove newline separation (join the lines above and below to one line as code text, single back tick formatted)

> There is a scenario ... this line here `*let* account_len: *usize* = 1 + 1 + (4 + *title*.len()) + (4 + *description*.len());` and the rent required ... check out [this blog post](https://dev.to/jacobcreech/how-to-change-account-size-on-solana-55b4).

sugg2: add it as a codeblock

> There is a scenario ... this line here:

```tsx
*let* account_len: *usize* = 1 + 1 + (4 + *title*.len()) + (4 + *description*.len());
```
> ...and the rent required ...